packages/loopring_v3/contracts/impl/libexchange/ExchangeData.sol

```diff
@@ -186,7 +186,7 @@ library ExchangeData
     }
     function FEE_BLOCK_FINE_START_TIME() internal pure returns (uint32) { return 6 hours; }
     function FEE_BLOCK_FINE_MAX_DURATION() internal pure returns (uint32) { return 6 hours; }
-    function MIN_GAS_TO_DISTRIBUTE_WITHDRAWALS() internal pure returns (uint32) { return 150000; }
+    function MIN_GAS_TO_DISTRIBUTE_WITHDRAWALS() internal pure returns (uint32) { return 30000; }
```

Comment:
I think the limit should be higher to be safe.

```
uint gasLimit = ExchangeData.MIN_GAS_TO_DISTRIBUTE_WITHDRAWALS();
while (totalNumWithdrawn < end && gasleft() >= gasLimit) {
  withdrawFromApprovedWithdrawal(...);
}
// Other stuff
```
In the worst case `withdrawFromApprovedWithdrawal` can cost about 100,000 gas now (`GAS_LIMIT_SEND_TOKENS()` + state reading/updating, though in most cases it will be less).

With `MIN_GAS_TO_DISTRIBUTE_WITHDRAWALS() == 30000` and `gasleft() == 30000` the loop would continue which means we could get out of gas in `withdrawFromApprovedWithdrawal` failing the whole transaction so all the successful calls to `withdrawFromApprovedWithdrawal` are reverted as well.

We also need some additional gas to exit the function after the loop when a fine needs to be paid:
```
S.loopring.burnExchangeStake(S.id, amountToBurn);
S.loopring.withdrawExchangeStake(S.id, msg.sender, amountToDistributer);
```

The current value of `MIN_GAS_TO_DISTRIBUTE_WITHDRAWALS()` of 150,000 gas should be safe to do a call to `withdrawFromApprovedWithdrawal` + successfully exit the function without running out of gas.

Reply:
@kongliangzhong

Reply:
Yes, set MIN_GAS_TO_DISTRIBUTE_WITHDRAWALS() to 150000 will be much safer.

`        while (totalNumWithdrawn < end && gasleft() >= gasLimit) {
            // Don't check the return value here, the withdrawal is allowed to fail.
            // The automatic token disribution by the operator is a best effort only.
            // The account owner can always manually withdraw without any limits.
            withdrawFromApprovedWithdrawal(
                S,
                blockIdx,
                withdrawBlock,
                totalNumWithdrawn,
                true
            );
            totalNumWithdrawn++;
        }
        withdrawBlock.numWithdrawalsDistributed = uint16(totalNumWithdrawn);
`
The parameter allowFailure in withdrawFromApprovedWithdrawal is always true, so the transaction will not fail if withdrawFromApprovedWithdrawal failed. So i think  MIN_GAS_TO_DISTRIBUTE_WITHDRAWALS can be set smaller.

Reply:
> Here the parameter `allowFailure` of withdrawFromApprovedWithdrawal is always true, so the transaction will not fail if withdrawFromApprovedWithdrawal failed. So i think MIN_GAS_TO_DISTRIBUTE_WITHDRAWALS can be set smaller.

The transaction will definitely fail when there's not enough gas left. `allowFailure` only works on the actual ETH/token transfer (and this will use up to `GAS_LIMIT_SEND_TOKENS` gas). Let's say there's `gasleft() == GAS_LIMIT_SEND_TOKENS` and you do an ETH transfer that costs `> GAS_LIMIT_SEND_TOKENS`. This transfer will fail, which is fine when `allowFailure ==  true`, but after the call we have no more gas left (`gasleft() == 0`)! So any logic needed to finish `withdrawFromApprovedWithdrawal` and/or `distributeWithdrawals` cannot be done, so the complete transaction will fail. 

Easy example: after the loop we do this:
`withdrawBlock.numWithdrawalsDistributed = uint16(totalNumWithdrawn);`
This alone will cost 5,000 or 20,000 gas, but we'd be completely out of gas so we can't even add 2 numbers together without running out of gas.

So to be safe `MIN_GAS_TO_DISTRIBUTE_WITHDRAWALS := maxGasCost(withdrawFromApprovedWithdrawal) + maxGasCost(distributeWithdrawals logic after the loop has ended)`